include/circt/Dialect/OM/Evaluator/Evaluator.h

```diff
@@ -328,6 +332,140 @@ struct TupleValue : EvaluatorValue {
+    return path;
+  }
+
+  /// Set the basepath which this path is relative to.
+  void setBasepath(const BasePathValue &basepath) {
+    assert(!isFullyEvaluated());
+    auto newPath = llvm::to_vector(basepath.path.getPath());
+    auto oldPath = path.getPath();
+    newPath.append(oldPath.begin(), oldPath.end());
+    path = PathAttr::get(path.getContext(), newPath);
+    markFullyEvaluated();
+  }
+
+  /// Finalize the evaluator value.
+  LogicalResult finalizeImpl() { return success(); }
+
+  /// Implement LLVM RTTI.
+  static bool classof(const EvaluatorValue *e) {
+    return e->getKind() == Kind::BasePath;
+  }
+
+private:
+  om::PathAttr path;
+};
+
+/// A Path value.
+struct PathValue : EvaluatorValue {
+  /// Create a path value representing a regular path.
+  PathValue(om::TargetKindAttr targetKind, om::PathAttr path, StringAttr module,
+            StringAttr ref, StringAttr field, Location loc)
+      : EvaluatorValue(loc.getContext(), Kind::Path, loc),
+        targetKind(targetKind), path(path), module(module), ref(ref),
+        field(field) {}
+
+  static PathValue getEmptyPath(Location loc) {
+    PathValue path(nullptr, nullptr, nullptr, nullptr, nullptr, loc);
+    path.markFullyEvaluated();
+    return path;
+  }
+
+  om::TargetKindAttr getTargetKind() const { return targetKind; }
+
+  om::PathAttr getPath() const { return path; }
+
+  StringAttr getModule() const { return module; }
+
+  StringAttr getRef() const { return ref; }
+
+  StringAttr getField() const { return field; }
+
+  StringAttr getAsString() const {
+    // If the module is null, then this is a path to a deleted object.
+    if (!targetKind)
+      return StringAttr::get(getContext(), "OMDeleted");
+    SmallString<64> result;
+    switch (targetKind.getValue()) {
+    case om::TargetKind::DontTouch:
+      result += "OMDontTouchedReferenceTarget";
+      break;
+    case om::TargetKind::Instance:
+      result += "OMInstanceTarget";
+      break;
+    case om::TargetKind::MemberInstance:
+      result += "OMMemberInstanceTarget";
+      break;
+    case om::TargetKind::MemberReference:
+      result += "OMMemberReferenceTarget";
+      break;
+    case om::TargetKind::Reference:
+      result += "OMReferenceTarget";
+      break;
+    }
+    result += ":~";
+    if (!path.getPath().empty())
+      result += path.getPath().front().module;
+    else
+      result += module.getValue();
+    result += '|';
+    for (const auto &elt : path) {
+      result += elt.module.getValue();
+      result += '/';
+      result += elt.instance.getValue();
+      result += ':';
+    }
+    if (!module.getValue().empty())
+      result += module.getValue();
+    if (!ref.getValue().empty()) {
+      result += '>';
+      result += ref.getValue();
+    }
+    if (!field.getValue().empty())
+      result += field.getValue();
+    return StringAttr::get(field.getContext(), result);
+  }
+
+  void setBasepath(const BasePathValue &basepath) {
+    assert(!isFullyEvaluated());
```

Comment:
nit: I think several of these longer method definitions like `setBasePath` and `getAsString` could be defined in the .cpp file to keep the header lightweight since they don't need any templates etc.

Reply:
Thanks, moved a bunch of functions over.